src/EventGridExtension/OutputBinding/EventGridAsyncCollector.cs

```diff
@@ -0,0 +1,42 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Azure.EventGrid;
+using Microsoft.Azure.EventGrid.Models;
+
+namespace Microsoft.Azure.WebJobs.Extensions.EventGrid
+{
+    public sealed class EventGridAsyncCollector : IAsyncCollector<EventGridEvent>
+    {
+        private readonly EventGridClient _client;
+        private readonly IList<EventGridEvent> _eventsToSend = new List<EventGridEvent>();
+        private readonly EventGridAttribute _attribute;
+
+        public EventGridAsyncCollector(EventGridAttribute attr)
+        {
+            _attribute = attr;
+
+            if (_client == null)
+            {
+                _client = new EventGridClient(new TopicCredentials(_attribute.SasKey));
+            }
+        }
+
+        public Task AddAsync(EventGridEvent item, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            _eventsToSend.Add(item);
+
+            return Task.CompletedTask;
+        }
+
+        public async Task FlushAsync(CancellationToken cancellationToken = default(CancellationToken))
+        {
+            if (_eventsToSend.Any())
```

Comment:
Many of our bindings don't currently do it, but we should be careful here we this is not currently thread safe. If `Add`/`AddAsync` are called while this is taking place, that will be modifying the collection your passing down to the SDK. We should guard against that here.

Reply:
took a stab at this with a `ManualResetEventSlim` object. lmk your thoughts